### PR TITLE
User bio

### PR DESCRIPTION
**Pull request by eduardostuart**
Allow users to add a short bio of 140 characters. (reference: #283)

1. Added User bio content into `user profile` page.
1. New migration file - bio column (144 characters, nullable)
1. Styles updated - add margins and max-width to bio text
1. Feature tests updated
1. `UpdateProfileRequest` updated. Added `sometimes|max:144` for bio field.
1.  Add boot event handlers to remove html tags from bio

## Preview

screen shot 2017-07-08 at 23 06 11
screen shot 2017-07-08 at 23 08 49
screen shot 2017-07-08 at 23 11 29

### CODE REVIEW

**Comment by driesvints** [comment]
@eduardostuart thank you for sending this in! I was already reviewing but saw you pushed some more commits. Request a review from me when you're done.

**Comment by eduardostuart** [comment]
@driesvints ok!

Should I also add `bio` field into registration? What do you think?

**Comment by eduardostuart** [comment]
@driesvints can you review?

File: app/Http/Requests/UpdateProfileRequest.php [hunk 1]
```diff
@@ -12,9 +12,15 @@ public function rules()
             'name' => 'required|max:255',
             'email' => 'required|email|max:255|unique:users,email,'.Auth::id(),
             'username' => 'required|max:255|unique:users,username,'.Auth::id(),
+            'bio' => 'sometimes|max:144',
```
Comment (driesvints) [inline app/Http/Requests/UpdateProfileRequest.php]: Increase this to 160. I also don't think we need to use `sometimes` here?

File: app/Http/Requests/UpdateProfileRequest.php [hunk 2]
```diff
@@ -12,9 +12,15 @@ public function rules()
...
         ];
     }
 
+    public function bio(): ?string
```
Comment (driesvints) [inline app/Http/Requests/UpdateProfileRequest.php]: Remove the nullable return type

File: app/Http/Requests/UpdateProfileRequest.php [hunk 3]
```diff
@@ -12,9 +12,15 @@ public function rules()
...
+    {
+        return $this->get('bio');
```
Comment (driesvints) [inline app/Http/Requests/UpdateProfileRequest.php]: Return a default value of an empty string here

File: app/Jobs/UpdateProfile.php [hunk 4]
```diff
@@ -29,6 +29,7 @@ public static function fromRequest(User $user, UpdateProfileRequest $request): s
             'name' => $request->name(),
             'email' => $request->email(),
             'username' => strtolower($request->username()),
+            'bio' => $request->bio(),
```
Comment (driesvints) [inline app/Jobs/UpdateProfile.php]: Replace this with `'bio' => trim(strip_tags($request->bio()),`

File: app/User.php [hunk 5]
```diff
@@ -38,13 +38,26 @@ class User extends Authenticatable
...
+        parent::boot();
+
+        $eventHandler = function ($model) {
+            $model->bio = ! $model->hasBio() ? null : trim(strip_tags($model->bio()));
+        };
+
+        static::creating($eventHandler);
+        static::updating($eventHandler);
+    }
+
```
Comment (driesvints) [inline app/User.php]: Remove this, we can do this in the job.

File: app/User.php [hunk 6]
```diff
@@ -141,6 +154,16 @@ public function isAdmin(): bool
         return $this->type() === self::ADMIN;
     }
 
+    public function bio(): ?string
```
Comment (driesvints) [inline app/User.php]: Remove the nullable type hint

File: app/User.php [hunk 7]
```diff
@@ -141,6 +154,16 @@ public function isAdmin(): bool
...
+    {
+        return $this->getAttributeValue('bio');
```
Comment (driesvints) [inline app/User.php]: Just return `$this->bio` here.

File: database/migrations/2017_07_08_202506_add_user_bio_column.php [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('users', function ($table) {
+            $table->string('bio', 140)->nullable();
```
Comment (driesvints) [inline database/migrations/2017_07_08_202506_add_user_bio_column.php]: Make this non-nullable. Increase this to 160 characters. Add a default value of an empty string.

File: database/migrations/2017_07_08_202506_add_user_bio_column.php [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
+    {
+        Schema::table('users', function ($table) {
+            $table->dropColumn('bio');
+        });
+    }
```
Comment (driesvints) [inline database/migrations/2017_07_08_202506_add_user_bio_column.php]: Remove the down migration.

File: resources/assets/sass/_profile.scss [hunk 10]
```diff
@@ -1,7 +1,13 @@
 .profile-user-info {
     padding-top: 25px;
     text-align: center;
+
+    .profile-user-bio {
+        margin:5px auto;
```
Comment (driesvints) [inline resources/assets/sass/_profile.scss]: Add a space after `:`

File: resources/views/users/_user_info.blade.php [hunk 11]
```diff
@@ -10,6 +10,11 @@
     @endcan
 
     <h2>{{ $user->name() }}</h2>
+    @if ($user->hasBio())
```
Comment (driesvints) [inline resources/views/users/_user_info.blade.php]: Add a white line above the if statement.

File: resources/views/users/settings/profile.blade.php [hunk 12]
```diff
@@ -41,6 +41,15 @@
                     </div>
                 @endFormGroup
 
+                @formGroup('bio')
+                    {!! Form::label('bio', null, ['class' => 'col-md-3 control-label']) !!}
+
+                    <div class="col-md-6">
+                        {!! Form::textarea('bio', Auth::user()->bio(), ['class' => 'form-control','rows'=>3]) !!}
```
Comment (driesvints) [inline resources/views/users/settings/profile.blade.php]: Replace this with

```
{!! Form::textarea('bio', Auth::user()->bio(), ['class' => 'form-control', 'rows' => 3, 'maxlength' => 160]) !!}
```

Comment (driesvints) [inline resources/views/users/settings/profile.blade.php]: Add a help text below the textarea and above the errors like so:

```
<span class="help-block">The user bio is limited to 160 characters.</span>
```

File: tests/Components/Models/UserTest.php [hunk 13]
```diff
@@ -37,6 +37,16 @@ public function it_can_return_the_amount_of_solutions_that_were_given()
...
+    /** @test */
+    public function it_cannot_contain_htmltags_in_bio()
+    {
+        $user = factory(User::class)->create([
+            'bio' => '<strong>my bio</strong>',
+        ]);
+
+        $this->assertEquals('my bio', $user->bio());
+    }
+
```
Comment (driesvints) [inline tests/Components/Models/UserTest.php]: Move this to a separate `UpdateProfileTest` like the other job tests: https://github.com/laravelio/portal/tree/master/tests/Components/Jobs

File: tests/CreatesUsers.php [hunk 14]
```diff
@@ -38,6 +38,7 @@ protected function createUser(array $attributes = []): User
             'email' => 'john@example.com',
             'password' => bcrypt('password'),
             'github_username' => 'johndoe',
+            'bio' => 'John Doe bio content',
```
Comment (driesvints) [inline tests/CreatesUsers.php]: This isn't needed here, remove it.

File: tests/Feature/SettingsTest.php [hunk 15]
```diff
@@ -46,11 +48,13 @@ public function users_cannot_choose_duplicate_usernames_or_email_addresses()
...
                 'email' => 'freek@example.com',
                 'username' => 'freekmurze',
+                'bio' => 'My short bio text',
             ])
             ->seePageIs('/settings')
             ->see('Something went wrong. Please review the fields below.')
             ->see('The email has already been taken.')
-            ->see('The username has already been taken.');
+            ->see('The username has already been taken.')
+            ->see('My short bio text');
```
Comment (driesvints) [inline tests/Feature/SettingsTest.php]: Remove this from this test.

File: tests/Feature/SettingsTest.php [hunk 16]
```diff
@@ -46,11 +48,13 @@ public function users_cannot_choose_duplicate_usernames_or_email_addresses()
                 'name' => 'Freek Murze',
                 'email' => 'freek@example.com',
                 'username' => 'freekmurze',
+                'bio' => 'My short bio text',
```
Comment (driesvints) [inline tests/Feature/SettingsTest.php]: Remove this from this test.

File: yarn.lock [hunk 17]
```diff
@@ -5125,6 +5112,18 @@ yargs@~3.10.0:
...
+  version "3.29.0"
+  resolved "https://registry.yarnpkg.com/yargs/-/yargs-3.29.0.tgz#1aab9660eae79d8b8f675bcaeeab6ee34c2cf69c"
+  dependencies:
+    camelcase "^1.2.1"
+    cliui "^3.0.3"
+    decamelize "^1.0.0"
+    os-locale "^1.4.0"
+    window-size "^0.1.2"
+    y18n "^3.2.0"
+
```
Comment (driesvints) [inline yarn.lock]: I'm not sure what went wrong here but please revert the changes to the `yarn.lock` file.

File: resources/views/users/_user_info.blade.php (hunk 11 above)
Comment (driesvints) [inline resources/views/users/_user_info.blade.php]: Replace this with `@if ($bio = $user->bio())` and use the `$bio` variable below.

File: app/User.php [hunk 18]
```diff
@@ -141,6 +154,16 @@ public function isAdmin(): bool
...
+    }
+
+    public function hasBio(): bool
+    {
+        return ! empty($this->bio());
+    }
```
Comment (driesvints) [inline app/User.php]: Remove this method.

**Comment by eduardostuart** [comment]
@driesvints  all done 😄

File: database/migrations/2017_07_08_202506_add_user_bio_column.php [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+
+    /**
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
+    {
+    }
```
Comment (driesvints) [inline database/migrations/2017_07_08_202506_add_user_bio_column.php]: You can remove the down method entirely :)

File: resources/views/users/_user_info.blade.php [hunk 20]
```diff
@@ -10,6 +10,12 @@
     @endcan
 
     <h2>{{ $user->name() }}</h2>
+    @if ($bio = $user->bio())
+
```
Comment (driesvints) [inline resources/views/users/_user_info.blade.php]: Remove this white line.

File: resources/views/users/_user_info.blade.php [hunk 21]
```diff
@@ -10,6 +10,12 @@
     @endcan
 
     <h2>{{ $user->name() }}</h2>
+    @if ($bio = $user->bio())
```
Comment (driesvints) [inline resources/views/users/_user_info.blade.php]: Add a white line above the if statement.

File: tests/Components/Jobs/UpdateProfileTest.php [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+    /** @test */
+    public function we_can_update_a_user_profile()
+    {
+        $user = $this->login(['name' => 'John Doe', 'email' => 'john@example.com', 'username' => 'johndoe']);
+
+        (new UpdateProfile($user, ['bio' => 'my bio', 'name' => 'John Doe Junior']))->handle();
+
+        $this->assertEquals('my bio', $user->bio());
+        $this->assertEquals('John Doe Junior', $user->name());
+    }
```
Comment (driesvints) [inline tests/Components/Jobs/UpdateProfileTest.php]: It's not needed to login the user. Since you're testing a component no "state" of the active user is needed. You can re-write this as follow:

```php
    /** @test */
    public function we_can_update_a_user_profile()
    {
        $user = (new UpdateProfile($this->createUser(), ['bio' => 'my bio', 'name' => 'John Doe Junior']))->handle();

        $this->assertEquals('my bio', $user->bio());
        $this->assertEquals('John Doe Junior', $user->name());
    }
```

Make sure that the `handle` method returns the updated user.

**Comment by eduardostuart** [comment]
@driesvints Done! 👍  Can you check?

**Comment by driesvints** [comment]
Thank you for your work on this! :)
